### PR TITLE
`x vendor`

### LINKED ISSUE #112391: Vendoring broken with stable rustc
I tried to enable vendoring in `config.toml` and then following the instructions given by `./x.py`.

I expected to see this happen: All dependencies are vendored.

Instead, this happened:

```
$ ./x.py build
error: vendoring required, but vendor directory does not exist.
       Run `cargo vendor --sync ./src/tools/cargo/Cargo.toml --sync ./src/tools/rust-analyzer/Cargo.toml --sync ./compiler/rustc_codegen_cranelift/Cargo.toml --sync ./src/bootstrap/Cargo.toml ` to initialize the vendor directory.
Alternatively, use the pre-vendored `rustc-src` dist component.
[...]
$ cargo vendor --sync ./src/tools/cargo/Cargo.toml --sync ./src/tools/rust-analyzer/Cargo.toml --sync ./compiler/rustc_codegen_cranelift/Cargo.toml --sync ./src/bootstrap/Cargo.toml
error: failed to load manifest for workspace member `/home/gh-bjorn3/rust/library/std`

Caused by:
  failed to parse manifest at `/home/gh-bjorn3/rust/library/std/Cargo.toml`

Caused by:
  the cargo feature `public-dependency` requires a nightly version of Cargo, but this is the `stable` channel
  See https://doc.rust-lang.org/book/appendix-07-nightly-rust.html for more information about Rust release channels.
  See https://doc.rust-lang.org/cargo/reference/unstable.html#public-dependency for more information about using this feature.
```

To work around this, I need to explicitly use a nightly cargo.

### Meta

Caused by https://github.com/rust-lang/rust/pull/111076 enabling the `public-dependency` unstable cargo feature.

### PR DESCRIPTION
**Pull request by onur-ozkan**
This PR implements `x vendor` on bootstrap; enabling dependency vendoring without the need for developers to have `cargo` installed on their system (previously, we suggested running `cargo vendor ...` but now we can accomplish the same task with `x vendor`).

In addition, fixes #112391 problem.

### CODE REVIEW

File: src/bootstrap/bootstrap.py [hunk 1]
```diff
@@ -1041,7 +1041,7 @@ def check_vendored_status(self):
                             "--sync ./compiler/rustc_codegen_gcc/Cargo.toml " \
                             "--sync ./src/bootstrap/Cargo.toml "
```
Comment (Mark-Simulacrum) [inline src/bootstrap/bootstrap.py:1042]: If we're going to have an x.py vendor, it seems like we might as well move this list into the source? Is there a reason to run it and not use all of these directories?

Comment (onur-ozkan) [reply src/bootstrap/bootstrap.py:1042]: Makes sense.

File: src/bootstrap/src/core/config/flags.rs [hunk 2]
```diff
@@ -456,6 +456,15 @@ Arguments:
         #[arg(long)]
         run: bool,
     },
+    /// Vendor dependencies
+    Vendor {
+        /// Additional `Cargo.toml` to sync and vendor
+        #[arg(long)]
+        sync: Vec<PathBuf>,
```
Comment (bjorn3) [inline src/bootstrap/src/core/config/flags.rs:463]: Why does this argument exist? Those extra workspaces wouldn't be used at all unless you modify the source code of bootstrap to make use of them, right? At that point you can also modify build_steps/vendor.rs to vendor your extra workspace.

**Pull request merged**